### PR TITLE
cellAudioOut: Report channel count based on settings + adjustments

### PR DESCRIPTION
**Pull request by capriots**
- cellAudioOutGetSoundAvailability() and cellAudioOutGetSoundAvailability2() will now return channel count based on downmix settings in the emulator.
- Pretend AC3 and DTS are available. There are games that only support surround over these formats and might internally downmix to stereo if only LPCM is available.
- For fs = 32KHz PS3 always returns 0.
- Removed error codes. PS3 doesn't appear to ever return those in these functions.

WipEout HD for example can now use its own internal stereo mode which sounds noticeably different from simply downmixing the surround output.

### CODE REVIEW

**Review by Megamouse — COMMENTED** [review]
One question: How often is this called?
audio_channel_downmix is a dynamic setting.

File: rpcs3/Emu/Cell/Modules/cellAudioOut.cpp [hunk 1]
```diff
@@ -32,11 +40,10 @@ error_code cellAudioOutGetSoundAvailability(u32 audioOut, u32 type, u32 fs, u32
 {
 	cellSysutil.warning("cellAudioOutGetSoundAvailability(audioOut=%d, type=%d, fs=0x%x, option=%d)", audioOut, type, fs, option);
 
-	s32 available = 8; // should be at least 2
+	s32 available = g_audio_out_chnum_id.at(g_cfg.audio.audio_channel_downmix);
 
 	switch (fs)
 	{
-	case CELL_AUDIO_OUT_FS_32KHZ:
```
[the remark this answers is not in the document]
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellAudioOut.cpp:39]: nevermind

File: rpcs3/Emu/Cell/Modules/cellAudioOut.cpp [hunk 2]
```diff
@@ -45,36 +52,45 @@ error_code cellAudioOutGetSoundAvailability(u32 audioOut, u32 type, u32 fs, u32
...
+		}
 
-	default: return CELL_AUDIO_OUT_ERROR_UNSUPPORTED_SOUND_MODE;
+	default: return not_an_error(0);
 	}
 
 	switch (audioOut)
 	{
-	case CELL_AUDIO_OUT_PRIMARY: return not_an_error(available);
-	case CELL_AUDIO_OUT_SECONDARY: return not_an_error(0);
```
[the remark this answers is not in the document]
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellAudioOut.cpp:63]: nevermind

File: rpcs3/Emu/Cell/Modules/cellAudioOut.cpp [hunk 3]
```diff
@@ -45,36 +52,45 @@ error_code cellAudioOutGetSoundAvailability(u32 audioOut, u32 type, u32 fs, u32
...
 
-	s32 available = 8; // should be at least 2
+	if (ch > g_audio_out_chnum_id.at(g_cfg.audio.audio_channel_downmix))
+	{
+		return not_an_error(0);
+	}
 
 	switch (fs)
 	{
-	case CELL_AUDIO_OUT_FS_32KHZ:
```
[the remark this answers is not in the document]
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellAudioOut.cpp:77]: nevermind

File: rpcs3/Emu/Cell/Modules/cellAudioOut.cpp [hunk 4]
```diff
@@ -83,36 +99,41 @@ error_code cellAudioOutGetSoundAvailability2(u32 audioOut, u32 type, u32 fs, u32
...
+		}
 
-	default: return CELL_AUDIO_OUT_ERROR_UNSUPPORTED_SOUND_MODE;
+	default: return not_an_error(0);
 	}
 
 	switch (audioOut)
 	{
-	case CELL_AUDIO_OUT_PRIMARY: return not_an_error(available);
-	case CELL_AUDIO_OUT_SECONDARY: return not_an_error(0);
```
[the remark this answers is not in the document]
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellAudioOut.cpp:110]: nevermind

**Comment by elad335** [comment]
`https://github.com/RPCS3/rpcs3/pull/12088#pullrequestreview-988429226`
I think it shouldn't be dynamic then.

**Comment by Megamouse** [comment]
I've actually added an alternative approach here:
#12112

**Comment by Megamouse** [comment]
@capriots thx for finding the bug.
I've added audio formats in the prior mentioned PR which reworks the dowmix and cellAudioOut logic.
Can you please test if it also fixes Wipeout HD AC3/DTS audio ?

**Comment by capriots** [comment]
Will do

**Pull request closed without merging**